junit-platform-engine/src/main/java/org/junit/platform/engine/support/descriptor/FileSource.java

```diff
@@ -55,27 +55,45 @@ public static FileSource from(File file) {
 	 * @param filePosition the position in the source file; may be {@code null}
 	 */
 	public static FileSource from(File file, @Nullable FilePosition filePosition) {
-		return new FileSource(file, filePosition);
+		return canonicalized(file, filePosition);
+	}
+
+	/**
+	 * Create a new {@code FileSource} from an existing instance but with a different
+	 * {@link FilePosition}. This avoids redundant canonical path resolution.
+	 *
+	 * @param source the existing {@code FileSource}; must not be {@code null}
+	 * @param filePosition the new {@code FilePosition}; may be {@code null}
+	 * @return a new {@code FileSource} with same file and updated position
+	 */
+	public static FileSource withPosition(FileSource source, @Nullable FilePosition filePosition) {
```

Comment:
This needs an `@API` annotation to mark the method as experimental with a `since = "1.14"` field.